### PR TITLE
Change how pointer casts work

### LINKED ISSUE #208: Be more careful when doing integer-to-pointer casts
Currently, miri will do integer-to-pointer casts my just loading the integer from memory as a pointer type. I implemented this before fully understanding the casting machinery and the role it plays in miri.

From what I understand now, the casting machinery e.g. guarantees that when I get a `Bytes(x)` somewhere of type `u32`, `x` will actually be in-bounds for `u32`.
We should, in the same vein, guarantee that pointers of reference type are non-NULL and aligned. Operations manipulating these references generally maintain that invariant, so casts should be the only place where we have to check.

### PR DESCRIPTION
**Pull request by RalfJung**
This re-implements int-ptr-casts as actual casts rather than sth. more transmute-like.

EDIT: Fixes #208.

### CODE REVIEW

**Comment by RalfJung** [comment]
I now also implemented an alignment check. This implements the strictest possible interpretation. See <https://internals.rust-lang.org/t/rules-for-alignment-and-non-nullness-of-references/5430> for a discussion of other possible interpretations.

In particular, this somewhat equates the two `reference_to_packed*` test cases. It makes little sense to keep them both around if already taking the reference (which happens before the two test cases differ) is UB. However, maybe it's worth keeping them around if we decide to change the rules later?

File: src/cast.rs [hunk 1]
```diff
@@ -94,8 +107,10 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
     fn cast_ptr(&self, ptr: PrimVal, ty: Ty<'tcx>) -> EvalResult<'tcx, PrimVal> {
         use rustc::ty::TypeVariants::*;
         match ty.sty {
-            TyRef(..) | TyRawPtr(_) | TyFnPtr(_) | TyInt(_) | TyUint(_) =>
+            // Casting to a reference or fn pointer is not permitted by rustc, no need to support it here.
```
Comment (oli-obk) [inline src/cast.rs]: They should probably be `=> bug!()` in that case, and not report a miri error.

File: src/eval_context.rs [hunk 2]
```diff
@@ -669,6 +670,26 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
                         bug!("attempted to take a reference to an enum downcast lvalue"),
                 };
 
+                // Check alignment and non-NULLness.
+                let (_, align) = self.size_and_align_of_dst(ty, val)?;
+                if ptr.is_ptr() {
```
Comment (oli-obk) [inline src/eval_context.rs]: `if let Ok(ptr) = ptr.to_ptr()`

Comment (oli-obk) [inline src/eval_context.rs]: or actually... match on the primval and get the inner parts. In the else path you do a `to_u128()` and in the main path a `to_ptr()`, I think a match would be clearer here.

File: src/eval_context.rs [hunk 3]
```diff
@@ -707,19 +728,9 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-                                // Alternatively, if the sizes are equal, try just reading at the target type
-                                .or_else(|err| {
-                                    let size = self.type_size(src_ty)?;
-                                    if size.is_some() && size == self.type_size(dest_ty)? {
-                                        self.value_to_primval(src, dest_ty)
-                                    } else {
-                                        Err(err)
-                                    }
-                                });
-                            self.write_value(Value::ByVal(dest_val?), dest, dest_ty)?;
```
Comment (oli-obk) [inline src/eval_context.rs]: 🎉

File: src/cast.rs (hunk 1 above)
Comment (RalfJung) [inline src/cast.rs]: Well, they are not really any different than all the other cases covered by the `_ => return Err(...)`.

**Comment by RalfJung** [comment]
It actually seems like there was no deep reason (yet?) to make ZST pointers aligned -- though there may be in the future: <https://internals.rust-lang.org/t/rules-for-alignment-and-non-nullness-of-references/5430/5>.

Let me know if you want me to keep or remove this check.

**Comment by oli-obk** [comment]
It would be fine with me to keep the stricter checking right now. Aren't here some alignment rules around empty arrays already?

**Comment by eddyb** [comment]
@oli-obk Yes, `[T; n]` has the alignment of `T` regardless of `n`.

**Comment by RalfJung** [comment]
> Yes, [T; n] has the alignment of T regardless of n.

The question is, when does this matter? AFAIK the only place alignment currently matters is loading from memory, because we tell LLVM that the loads are aligned.

**Comment by eddyb** [comment]
@RalfJung LLVM assumes that pointers are aligned for their pointee, and will optimize based on that, e.g. if you try to do the bit twiddling to align a pointer to that alignment it knows, you get a noop.

**Comment by RalfJung** [comment]
@eddyb But that doesn't apply to Rust's raw pointers, does it? I mean, the following program is considered safe by Rust:
```rust
let x = 13 as *const u32;
```
There seems to be a dedicated attribute to tell LLVM when to assume correct alignment:
> `align <n>`
    This indicates that the pointer value may be assumed by the optimizer to have the specified alignment.
    Note that this attribute has additional semantics when combined with the byval attribute.

**Comment by oli-obk** [comment]
But if the pointer comes from a reference, won't the optimizer be able to infer this information onto the raw pointer?

**Comment by RalfJung** [comment]
I would think so? The point here is to determine whether just *creating* an unaligned pointer or reference is UB. The check I added to miri goes half-way to implement a policy that says "yes for references, no for pointers". (Half-way because one can still create unaligned references via `transmute`... same for invalid booleans, btw)

**Comment by eddyb** [comment]
@RalfJung Well... sure... that's safe in Rust and I don't know how to cause UB with it *in safe code* because all that happens AFAIK is you get weird pointer arithmetic. But it will break unsafe code.

**Comment by eddyb** [comment]
> There seems to be a dedicated attribute to tell LLVM when to assume correct alignment:

Mno, in LLVM *everything* is aligned by default, those attributes *change* the alignment from LLVM's understanding of ABI alignment given the type (in this case, it's the pointee type).

**Comment by RalfJung** [comment]
@eddyb If that weird pointer arithmetic is observable in safe Rust, then I'd consider this a bug. Safe Rust should have the "obvious" semantics. It's bad enough that raw pointer comparison is permitted, but there we can at least get away with "result is non-deterministic", which is not-too-bad -- it's just a boolean, after all.

**Pull request merged**